### DISCUSSION TITLE
Very poor performance after 3.7.8 to 4.0.5 upgrade

### DISCUSSION THREAD

**Discussion by Kage1** [netbox-community/netbox#16684] [Help Wanted!] [answered]
### Deployment Type

Self-hosted

### NetBox Version

v4.0.5

### Python Version

3.10

### Steps to Reproduce

1. Backup database and took VM snap shot of current 3.7.8 version
2. Pull the current netbox git repository against master
3. Run ./upgrade.sh from the /etc/netbox/ install directory
4. Upgrade.sh runs with out an error
5. Restarted the netbox and netbox-rq services
6. Flush cookies/cache and open a new tab to the netbox instance. Web interface is extremely slow, 30+ seconds to navigtate to any page in the system. 3.7.8 version was very snappy, 1 sec or less load times for just about everything.

OS version: Ubuntu 22.04.4 LTS x64
Netbox version: 4.0.5
Redis version:  Redis 6.0.16
Postgre db version: PostgreSQL 14.12
Postgre db size: ~ 51 MB

I checked the NB, Redis, and Postgre logs and there are no issues being reported. htop does not show any abnormal process activity, e.g. the python3 environments running are not spiking the processor etc. and memory utilizing is approx 80% free.

### Expected Behavior

Similar performance as before.

### Observed Behavior

Extremely slow performance to the point of being unusable.

**Comment by jeremystretch** [comment]
Are you using Safari? If so, see #16447.

**Reply by Kage1** [reply]
I'm using Firefox 123.01 and tried with Edge.

**Comment by Kage1** [comment]
It may look like the reindex didnt run... when I try to run it by hand I get:

```
/opt/netbox# python3 netbox/manage.py reindex
Traceback (most recent call last):
  File "/opt/netbox/netbox/manage.py", line 8, in <module>
    from django.core.management import execute_from_command_line
ModuleNotFoundError: No module named 'django'
```

**Reply by Kage1** [reply]
Finally figured out I needed to run the manage.py through the virtual python environment:

`./venv/bin/python3 ./netbox/manage.py reindex`

The reindex ran and completed w/o an issue, but the performance is still abysmal.

**Comment by markkuleinio** [comment]
What does the browser devtools (or whatever they are called) show, where is the time spent?

**Reply by Kage1** [reply]
During a reload there is no html loaded... from the surface it appears there's is something going on in the backend that is causing the lag. Once the HTML render's it's pretty much immediate.

**Reply by Kage1** [reply]
Did some timing. From clicking on a menu link in Netbox till data is displayed it's anywhere from 20 - 30 seconds. It doesn't matter if there is nothing in that section (e.g. Circuits) or selecting IP Addresses which loads the first 50 in less than a second. It also doesn't matter if I have been to that section before, clicking between Circuits and IP Addresses always takes 20 - 30 seconds.

I do have one custom link, but even with that disabled it didn't change anything.

**Comment by julianstolp** [comment]
Have you fixed this problem somehow? I also have this problem coming from 3.7.8 to 4.0.5. Using Edge Browser

**Reply by julianstolp** [reply]
![image](https://github.com/netbox-community/netbox/assets/41152397/fa249ccd-fb30-46ac-9d25-094248dd5dba) 

Switching from gunicorn to uwsgi does not fix this problem.

**Reply by Kage1** [reply]
No, I had to roll back to my VM snapshot on version 3.7.8. The performance was was so poor it was to the point of unusable.

**Reply by julianstolp** [reply]
**I did a downgrade to 4.0.3 which fixes this issue for me.** @jeremystretch can you please look into this as this affects more people?

**Reply by jeremystretch** [reply]
I'd be happy to look into any specific problem someone can point me to. Right now it doesn't appear that anyone can provide steps to reproduce the issue.

**Reply by Kage1** [reply]
Do you have any logs we can pull, tests you'd like us to do?

From the screenshot you can see the browser was waiting on the server for 30+ seconds for a response.

Per @julianstolp rolling back to 4.0.3 has restored the performance.

What updates have made from 4.0.3 to now that could be causing it?

**Reply by julianstolp** [reply]
@Kage1 have you tried upgrading to 4.0.3 to confirm my observations? I tested this with to different servers, same behavior.

**v4.0.5**
```
uwsgi[597238]: [pid: 597238|app: 0|req: 1/1] <IP REMOVED> () {80 vars in 1741 bytes} [Mon Jun 24 14:39:10 2024] GET / => generated
123251 bytes in 27431 msecs (HTTP/1.0 200) 10 headers in 477 bytes (1 switches on core 0)
```

**v4.0.3**
```
uwsgi[597904]: [pid: 597904|app: 0|req: 2/7] <IP REMOVED> () {80 vars in 1745 bytes} [Mon Jun 24 14:49:02 2024] GET / => generated
123233 bytes in 174 msecs (HTTP/1.0 200) 10 headers in 477 bytes (1 switches on core 1)
```

**Reply by Kage1** [reply]
> @Kage1 have you tried upgrading to 4.0.3 to confirm my observations? I tested this with to different servers, same behavior.

Not yet we are in the middle of some cutovers so we need NB up.

If you go to 4.0.3 then to 4.0.5 you still see the slow response?

**Reply by julianstolp** [reply]
Unfortunately, downgrading to v4.0.3 and then upgrading to v4.0.5 does not solve the problem. I have tried the following things so far, without success:

- Completely empty database
- Switch between uwsgi and gunicorn
- Using the development server python3 manage.py runserver 0.0.0.0:8000 --insecure
- Django debug toolbar is also very slow, shows long request time

In version 4.0.3 everything works perfectly.

**Reply by markkuleinio** [reply]
> Completely empty database

Now that's interesting

**Reply by julianstolp** [reply]
Problem also disappeared with v4.0.6. Still no idea why it happened.

**Reply by markkuleinio** [reply]
> https://github.com/netbox-community/netbox/issues/16454 - Address DNS lookup bug in django-debug-toolbar

?

**Reply by julianstolp** [reply]
Thanks, just had a quick look at the patch notes. Yes, downgrading the debug toolbar from 4.4.2 to 4.3.0 in version 4.0.5 fixes the problem.

**Reply by Kage1** [reply, marked as answer]
#16506 "Roll back django-debug-toolbar version to avoid DNS lookup bug" was merged for v4.0.6. I still need to test it, but @julianstolp confirmed that 4.0.6 fixed the issue for him.

**Reply by Kage1** [reply]
Wrapping up, I was able to replicate our prod instance into our lab. I upgraded it to 4.0.6 and the performance it back to normal. I can confirm that patch #16506 did indeed fix this issue.

**Comment by Delta1977** [comment]
Same here, with Edge and Chrome 
Coming from 3.0.1 updating through every Version up to 4.0.5.

The Performance Impact  appears with v4

**Reply by julianstolp** [reply]
Can you test with version 4.0.3?
